redis/src/main/scala/zio/redis/Interpreter.scala

```diff
@@ -1,135 +1,236 @@
+          .make[RedisError, RespValue]
+          .flatMap(promise => reqQueue.offer(Request(command, promise)) *> promise.await)
+
+      private def sendNext(out: Chunk[Byte] => IO[IOException, Unit]): IO[RedisError, Unit] =
+        reqQueue.takeBetween(1, Int.MaxValue).flatMap { reqs =>
+          val bytes = Chunk.fromIterable(reqs).flatMap(req => RespValue.Array(req.command).serialize)
+          out(bytes)
+            .mapError(RedisError.IOError)
+            .tapError(e => IO.foreach(reqs)(_.promise.fail(e)))
+            .zipLeft(IO.foreach(reqs)(req => resQueue.offer(req.promise)))
+        }
+
+      private def runReceive(inStream: Stream[IOException, Byte]): IO[RedisError, Unit] =
+        inStream
+          .mapError(RedisError.IOError)
+          .transduce(RespValue.deserialize.toTransducer)
+          .foreach(response => resQueue.take.flatMap(_.succeed(response)))
+
+      /**
+       * Opens a connection to the server and launches send and receive operations.
+       * All failures are retried by opening a new connection.
+       * Only exits by interruption or defect.
+       */
+      def run: IO[RedisError, Unit] =
+        byteStream
+          .mapError(RedisError.IOError)
+          .use { rwBytes =>
+            sendNext(rwBytes.write).forever race runReceive(rwBytes.read)
+          }
+          .tapError { e =>
+            IO.effectTotal(println(s"Reconnecting due to error: $e")) *>
+              resQueue.takeAll.flatMap(IO.foreach(_)(_.fail(e)))
+          }
+          .retryWhile(Function.const(true))
+          .tapCause(c => IO.effectTotal(Console.err.println(s"Executor exiting: $c")))
+
     }
 
-    def live(host: String, port: Int): ZLayer[Blocking, IOException, RedisExecutor] =
+    def live: ZLayer[ByteStream, RedisError.IOError, RedisExecutor] =
       ZLayer.fromServiceManaged { env =>
         for {
-          connection <- connect(host, port)
-          _          <- env.blocking(connection.receive.forever).forkManaged
-        } yield new Service {
-          def execute(command: Chunk[String]): IO[RedisError, String] = connection.send(command).flatMap(_.await)
-        }
+          reqQueue <- Queue.bounded[Request](requestQueueSize).toManaged_
+          resQueue <- Queue.unbounded[Promise[RedisError, RespValue]].toManaged_
+          live      = new Live(reqQueue, resQueue, env.connect)
+          _        <- live.run.forkManaged
+        } yield live
       }
 
-    private[this] def connect(host: String, port: Int): Managed[IOException, Connection] = {
-      val makeChannel =
-        IO {
-          val channel = SocketChannel.open(new InetSocketAddress(host, port))
-          channel.configureBlocking(false)
-          channel.setOption(StandardSocketOptions.SO_KEEPALIVE, Boolean.box(true))
-          channel.setOption(StandardSocketOptions.TCP_NODELAY, Boolean.box(true))
-          channel
-        }
+    def liveServer(address: SocketAddress): Layer[RedisError.IOError, RedisExecutor] =
```

Comment:
I would drop this.

Reply:
I think it's useful, sometimes it's good to have control over the address resolution. Also, it is technically _possible_ for a Java client to connect to Redis over a Unix domain socket (with an extra library), which would require the `SocketAddress` API to work.

Reply:
Hm, ok. Let's keep it, but please rename to `live`.